pkg/build/registry/buildconfig/etcd/etcd.go

```diff
@@ -45,5 +62,132 @@ func NewREST(s storage.Interface) *REST {
+	}
+
+	// otherwise, find the jenkins SA specified (is that missing) and get roles it has in this namespace
+	roleBindings, err := r.roleBindingClient.RoleBindings(bc.Namespace).List(kapi.ListOptions{})
+	if err != nil {
+		return nil, kapierrors.NewInternalError(err)
+	}
+	foundEdit := false
+	for _, roleBinding := range roleBindings.Items {
+		if len(roleBinding.RoleRef.Namespace) != 0 || roleBinding.RoleRef.Name != "edit" {
+			continue
+		}
+		for _, subject := range roleBinding.Subjects {
+			if subject.Kind == "ServiceAccount" && subject.Name == "jenkins" && subject.Namespace == bc.Namespace {
+				foundEdit = true
+			}
+		}
+	}
+
+	// we didn't find the role we wanted, try to add that role
+	if !foundEdit {
+		addRole := &policy.RoleModificationOptions{
+			RoleName: "edit",
+			RoleBindingAccessor: &impersonatingRoleBindingAccessor{
+				user:              user,
+				bindingNamespace:  bc.Namespace,
+				roleBindingClient: r.roleBindingClient,
+				restClient:        r.restClient,
+			},
+			Subjects: []kapi.ObjectReference{{Kind: "ServiceAccount", Namespace: bc.Namespace, Name: "jenkins"}},
+		}
+
+		if err := addRole.AddRole(); err != nil {
+			if kapierrors.IsForbidden(err) {
+				return nil, kapierrors.NewInternalError(fmt.Errorf("need to bind jenkins to edit: `oc policy add-role-to-user edit -z jenkins`"))
+			}
+			return nil, kapierrors.NewInternalError(err)
+		}
+	}
+
+	return r.Etcd.Create(ctx, obj)
+}
+
+type impersonatingRoleBindingAccessor struct {
+	user             user.Info
+	bindingNamespace string
+
+	roleBindingClient oclient.RoleBindingsNamespacer
+
+	// TODO pretty this up
+	restClient *restclient.RESTClient
+}
+
+func (a impersonatingRoleBindingAccessor) GetExistingRoleBindingsForRole(roleNamespace, role string) ([]*authorizationapi.RoleBinding, error) {
+	existingBindings, err := a.roleBindingClient.RoleBindings(a.bindingNamespace).List(kapi.ListOptions{})
+	if err != nil && !kapierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	ret := make([]*authorizationapi.RoleBinding, 0)
+	// see if we can find an existing binding that points to the role in question.
+	for i := range existingBindings.Items {
+		currBinding := &existingBindings.Items[i]
+		if currBinding.RoleRef.Namespace == roleNamespace && currBinding.RoleRef.Name == role {
+			t := currBinding
+			ret = append(ret, t)
+		}
+	}
+
+	return ret, nil
+}
+
+func (a impersonatingRoleBindingAccessor) GetExistingRoleBindingNames() (*sets.String, error) {
+	roleBindings, err := a.roleBindingClient.RoleBindings(a.bindingNamespace).List(kapi.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &sets.String{}
+	for _, currBinding := range roleBindings.Items {
+		ret.Insert(currBinding.Name)
+	}
+
+	return ret, nil
+}
+
+func (a impersonatingRoleBindingAccessor) UpdateRoleBinding(binding *authorizationapi.RoleBinding) error {
+	request := a.restClient.Put()
+	request = request.SetHeader(authenticationapi.ImpersonateUserHeader, a.user.GetName())
```

Comment:
This is the bit that makes this indirection instead of escalation.  It's seamless for those with power, its impossible for those without.